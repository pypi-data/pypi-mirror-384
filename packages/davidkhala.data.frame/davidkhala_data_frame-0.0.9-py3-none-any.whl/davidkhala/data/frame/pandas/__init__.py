import pandas as pd
def upsert(df, *primary_keys: str, record: dict):
    index_keys = df.index.names if isinstance(df.index, pd.MultiIndex) else [df.index.name]
    condition = True

    for key in primary_keys:
        if key in df.columns:
            condition &= (df[key] == record[key])
        elif key in index_keys:
            condition &= (df.index.get_level_values(key) == record[key])
        else:
            raise KeyError(f"'{key}' not found in either columns or index")

    match_indices = df[condition].index

    if not match_indices.empty:
        for col, value in record.items():
            if col in df.columns:
                df.loc[match_indices, col] = value
    else:
        df = pd.concat([df, pd.DataFrame([record])], ignore_index=True)

    return df