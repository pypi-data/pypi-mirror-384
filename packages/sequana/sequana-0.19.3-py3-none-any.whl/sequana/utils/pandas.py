#
#  This file is part of Sequana software
#
#  Copyright (c) 2016 - Sequana Development Team
#
#  File author(s):
#      Thomas Cokelaer <thomas.cokelaer@pasteur.fr>
#
#  Distributed under the terms of the 3-clause BSD license.
#  The full license is in the LICENSE file, distributed with this software.
#
#  website: https://github.com/sequana/sequana
#  documentation: http://sequana.readthedocs.io
#
##############################################################################
"""Simple utilities for pandas"""

from sequana.lazy import pandas as pd


class PandasReader:
    def __init__(self, filename, sep="\t", columns=None, **kwargs):
        try:
            self.df = pd.read_csv(filename, sep=sep, **kwargs)
        except pd.errors.EmptyDataError:
            self.df = pd.DataFrame(columns=columns)

        if columns:
            self.df.columns = columns

        # If there is a header, let us strip it from spaces
        try:
            self.df.columns = [x.strip() for x in self.df.columns]
        except:
            pass

        # let us strip strings from spaces
        for x in self.df.columns:
            try:
                self.df[x] = self.df[x].str.strip()
            except:
                pass
