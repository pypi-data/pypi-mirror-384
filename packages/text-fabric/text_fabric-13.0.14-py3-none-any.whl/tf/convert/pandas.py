"""
# Export a TF dataset to a `pandas` data frame.

There is a natural mapping of a TF dataset with its nodes, edges and features to a
rectangular data frame with rows and columns:

*   the *nodes* correspond to *rows*;
*   the node *features* correspond to *columns*;
*   the *value* of a feature for a node is in the row that corresponds with the node
    and the column that corresponds with the feature.
*   the *edge* features correspond to columns, in that column you find for each row
    the nodes where edges arrive, i.e. the edges from the node that correspond with
    the row.

We also write the data that says which nodes are contained in which other nodes.
To each row we add the following columns:

*   for each node type, except the slot type, there is a column with named
    `in_nodeType`, that contains the node of the smallest object that
    contains the node of the row;

We compose the big table and save it as a tab delimited file.
This temporary result can be processed by `R` and `pandas`.

It turns out that for this size of the data `pandas` is a bit
quicker than R. It is also more Pythonic, which is a pro if you use other Python
programs, such as TF, to process the same data.

# Examples

*   [BHSA](https://nbviewer.org/github/ETCBC/bhsa/blob/master/tutorial/export.ipynb)
*   [Moby Dick](https://nbviewer.org/github/annotation/mobydick/blob/main/tutorial/exportPandas.ipynb)
*   [Ferdinand Huyck](https://nbviewer.org/github/CLARIAH/wp6-ferdinandhuyck/blob/main/tutorial/export.ipynb)
"""

from ..capable import CheckImport
from ..parameters import OTYPE, OSLOTS
from ..core.files import TEMP_DIR, fileOpen, unexpanduser as ux, expandDir, dirMake
from ..core.helpers import fitemize, pandasEsc, PANDAS_QUOTE, PANDAS_ESCAPE


HELP = """
Transforms TF dataset into `pandas`
"""

INT = "Int64"
STR = "str"
NA = [""]


def exportPandas(app, inTypes=None, exportDir=None):
    """Export a currently loaded TF dataset to `pandas`.

    The function proceeds by first producing a TSV file as an intermediate result.
    This is usually too big for GitHub, to it is produced in a `/_temp` directory
    that is usually in the `.gitignore` of the repo.

    This file serves as the basis for the export to a `pandas` data frame.

    !!! hint "R"
        You can import this file in other programs as well, e.g.
        [R](https://www.r-project.org)

    !!! note "Quotation, newlines, tabs, backslashes and escaping"
        If the data as it comes from TF contains newlines or tabs or
        double quotes, we put them escaped into the TSV, as follows:

        *   *newline* becomes *backslash* plus `n`;
        *   *tab* becomes a single space;
        *   *double quote* becomes *Control-A* plus *double quote*;
        *   *backslash* remains *backslash*.

        In this way, the TSV file is not disturbed by non-delimiting tabs, i.e.
        tabs that are part of the content of a field. No field will contain a tab!

        Also, no field will contain a newline, so the lines are not disturbed by
        newlines that are part of the content of a field. No field will contain a
        newline!

        Double quotes in a TSV file might pose a problem. Several programs interpret
        double quotes as a way to include tabs and newlines in the content of a field,
        especially if the quote occurs at the beginning of a field.
        That's why we escape it by putting a character in front of it that is very
        unlikely to occur in the text of a corpus: Ctrl A, which is ASCII character 1.

        Backslashes are no problem, but programs might interpret them in a special
        way in combination with specific following characters.

        Now what happens to these characters when `pandas` reads the file?

        We instruct the `pandas` table reading function to use the Control-A as
        escape char and the double quote as quote char.

        **Backslash**

        `pandas` has two special behaviours:

        *   *backslash* `n` becomes a *newline*;
        *   *backslash* *backslash* becomes a single *backslash*.

        This is almost what we want: the newline behaviour is desired; the
        reducing of backslashes not, but we leave it as it is.

        **Double quote**

        *Ctrl-A* plus *double quote* becomes *double quote*.

        That is exactly what we want.

    Parameters
    ----------
    app: object
        A `tf.advanced.app.App` object that represent a loaded corpus, together with
        all its loaded data modules.
    inTypes: string | iterable, optional None
        A bunch of node types for which columns should be made that contain nodes
        in which the row node is contained.
        If `None`, all node types will have such columns. But for certain TEI corpora
        this might lead to overly many columns.
        So, if you specify `""` or `{}`, there will only be columns for sectional
        node types.
        But you can also specify the list of such node types explicitly.
        In all cases, there will be columns for sectional node types.
    exportDir: string, optional None
        The directory to which the `pandas` file will be exported.
        If `None`, it is the `/pandas` directory in the repo of the app.
    """
    CI = CheckImport("pandas", "pyarrow")
    if CI.importOK(hint=True):
        (pandas, pyarrow) = CI.importGet()
    else:
        return

    api = app.api
    Eall = api.Eall
    Fall = api.Fall
    Es = api.Es
    Fs = api.Fs
    F = api.F
    N = api.N
    L = api.L
    T = api.T
    TF = api.TF

    app.indent(reset=True)

    sectionTypes = T.sectionTypes
    sectionFeats = T.sectionFeats

    sectionTypeSet = set(sectionTypes)
    sectionFeatIndex = {}

    for i, f in enumerate(sectionFeats):
        sectionFeatIndex[f] = i

    skipFeatures = {f for f in Fall() + Eall() if "@" in f}

    textFeatures = set()
    for textFormatSpec in TF.cformats.values():
        for featGroup in textFormatSpec[2]:
            for feat in featGroup[0]:
                textFeatures.add(feat)
    textFeatures = sorted(textFeatures)

    inTypes = [
        t
        for t in (F.otype.all if inTypes is None else fitemize(inTypes))
        if t not in sectionTypes
    ]
    edgeFeatures = sorted(set(Eall()) - {OSLOTS} - skipFeatures)
    nodeFeatures = sorted(set(Fall()) - {OTYPE} - set(textFeatures) - skipFeatures)

    dtype = dict(nd=INT, otype=STR)

    for f in sectionTypes:
        dtype[f"in_{f}"] = INT

    for f in nodeFeatures:
        dtype[f] = INT if Fs(f).meta["valueType"] == "int" else STR

    naValues = dict((x, set() if dtype[x] == STR else {""}) for x in dtype)

    baseDir = f"{app.repoLocation}"
    tempDir = f"{baseDir}/{TEMP_DIR}"
    if exportDir is None:
        exportDir = f"{baseDir}/pandas"
    else:
        exportDir = exportDir
        exportDir = expandDir(app, exportDir)

    dirMake(tempDir)
    dirMake(exportDir)

    tableFile = f"{tempDir}/data-{app.version}.tsv"
    tableFilePd = f"{exportDir}/data-{app.version}.pd"

    chunkSize = max((100, int(round(F.otype.maxNode / 20))))

    app.info("Create tsv file ...")
    app.indent(level=True, reset=True)

    with fileOpen(tableFile, mode="w") as hr:
        cells = (
            "nd",
            "otype",
            *textFeatures,
            *[f"in_{x}" for x in sectionTypes],
            *[f"in_{x}" for x in inTypes],
            *edgeFeatures,
            *nodeFeatures,
        )
        hr.write("\t".join(cells) + "\n")
        i = 0
        s = 0
        perc = 0

        for n in N.walk():
            nType = F.otype.v(n)
            textValues = [pandasEsc(str(Fs(f).v(n) or "")) for f in textFeatures]
            sectionNodes = [
                n if nType == section else (L.u(n, otype=section) or NA)[0]
                for section in sectionTypes
            ]
            inValues = [(L.u(n, otype=inType) or NA)[0] for inType in inTypes]
            edgeValues = [pandasEsc(str((Es(f).f(n) or NA)[0])) for f in edgeFeatures]
            nodeValues = [
                pandasEsc(
                    str(
                        (Fs(f).v(sectionNodes[sectionFeatIndex[f]]) or NA[0])
                        if f in sectionFeatIndex and nType in sectionTypeSet
                        else Fs(f).v(n) or ""
                    )
                )
                for f in nodeFeatures
            ]
            cells = (
                str(n),
                F.otype.v(n),
                *textValues,
                *[str(x) for x in sectionNodes],
                *[str(x) for x in inValues],
                *edgeValues,
                *nodeValues,
            )
            hr.write("\t".join(cells).replace("\n", "\\n") + "\n")
            i += 1
            s += 1
            if s == chunkSize:
                s = 0
                perc = int(round(i * 100 / F.otype.maxNode))
                app.info(f"{perc:>3}% {i:>7} nodes written")

    app.info(f"{perc:>3}% {i:>7} nodes written and done")
    app.indent(level=False)

    app.info(f"TSV file is {ux(tableFile)}")

    with fileOpen(tableFile, mode="r") as hr:
        rows = 0
        chars = 0
        columns = 0
        for i, line in enumerate(hr):
            if i == 0:
                columns = line.split("\t")
                app.info(f"Columns {len(columns)}:")
                for col in columns:
                    app.info(f"\t{col}")
            rows += 1
            chars += len(line)
    app.info(f"\t{rows} rows")
    app.info(f"\t{chars} characters")

    app.info("Importing into Pandas ...")
    app.indent(level=True, reset=True)
    app.info("Reading tsv file ...")

    dataFrame = pandas.read_table(
        tableFile,
        delimiter="\t",
        quotechar=PANDAS_QUOTE.encode("utf-8"),
        escapechar=PANDAS_ESCAPE.encode("utf-8"),
        doublequote=False,
        low_memory=False,
        encoding="utf8",
        keep_default_na=False,
        na_values=naValues,
        dtype=dtype,
    )
    app.info("Done. Size = {}".format(dataFrame.size))

    app.info("Saving as Parquet file ...")

    dataFrame.to_parquet(tableFilePd, engine="pyarrow")
    app.info("Saved")
    app.indent(level=False)
    app.info(f"PD  in {ux(tableFilePd)}")
