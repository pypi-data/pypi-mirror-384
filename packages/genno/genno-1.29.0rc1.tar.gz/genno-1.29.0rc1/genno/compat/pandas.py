import logging
from contextlib import contextmanager
from functools import lru_cache

from packaging.version import Version, parse

log = logging.getLogger(__name__)


@lru_cache
def version() -> Version:
    import pandas

    return parse(pandas.__version__)


@contextmanager
def disable_copy_on_write(name):
    """Context manager to disable Pandas :ref:`pandas:copy_on_write`.

    A message is logged with level :any:`logging.DEBUG` if the setting is changed.
    """
    import pandas

    stored = pandas.options.mode.copy_on_write
    override_value = "warn" if version() >= Version("2.2.0") else False

    try:
        if stored is True:
            log.debug(f"Override pandas.mode.options.copy_on_write = True for {name}")
            pandas.options.mode.copy_on_write = override_value
        yield
    finally:
        pandas.options.mode.copy_on_write = stored


@lru_cache
def handles_parquet_attrs() -> bool:
    """Return :any:`True` if :mod:`pandas` can read/write attrs to/from Parquet files.

    If not, a message is logged.
    """
    if version() < Version("2.1.0"):
        log.info(
            f"Pandas {version()!s} < 2.1.0 cannot read/write Quantity.attrs "
            f"to/from Parquet; {__name__} will use pickle from the standard library"
        )
        return False
    else:
        return True
