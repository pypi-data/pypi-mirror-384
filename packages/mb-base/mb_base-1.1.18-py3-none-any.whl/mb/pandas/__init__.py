from mb_pandas import *
from pandas import *
