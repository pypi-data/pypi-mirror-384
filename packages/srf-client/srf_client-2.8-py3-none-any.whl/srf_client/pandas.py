from __future__ import annotations

import io
import math
from datetime import timedelta, timezone
from functools import partial, reduce
from typing import Any, Callable, Collection, List, Optional, Union, cast, \
    NamedTuple, MutableMapping

import pandas as pd

from .model import ChargerTransaction, Leg, Trip

DF_BiFunc = Callable[[pd.DataFrame, pd.DataFrame], pd.DataFrame]


def to_numeric(v: str) -> int | float:
    """Convert a string to a number, coercing errors into NaN."""
    try:
        return int(v)
    except ValueError:
        try:
            return float(v)
        except ValueError:
            return math.nan


class Data(NamedTuple):
    """Alternate shape for leg data."""

    index: List[int]
    data: List[tuple]


def get_data_frame(self: Union[Leg, Trip],
                   data_types: Union[str, Collection[str]],
                   resolution: Union[str, timedelta, None] = None,
                   conversion: Optional[Callable[[str], Any]] = to_numeric,
                   include_can_source: bool = None
                   ) -> pd.DataFrame:
    """
    Return available measurements as a time-series :code:`DataFrame`.

    :code:`Series` will be named :code:`{TYPE}_{FIELD}` in the order given
    by :code:`data_types`.

    Data will be interpolated to the specified resolution.

    :param data_types: Data types to fetch.
    :param resolution: Target resolution. If :code:`None` (the default) then
        it will use the approximate resolution of the first requested type.
    :param conversion: Conversion to apply to all values. If :code:`None`
        then all values will be strings.
    :param include_can_source: Include and partition by multiple CAN sources.

    .. versionchanged:: 2.3
       Added ``include_can_source`` parameter.
    """
    if isinstance(data_types, str):
        data_types = [data_types]

    collected: MutableMapping[str, Optional[Data]] = {
        dt: Data([], []) for dt in data_types
    }
    for m in self.get_data(include=data_types):
        collected[m.type].index.append(int(m.timestamp))
        row = m.data.split(',')
        if conversion is not None:
            if include_can_source:
                row[1:] = [conversion(v) for v in row[1:]]
            else:
                row = (conversion(v) for v in row)
        collected[m.type].data.append(tuple(row))

    to_dt = partial(pd.to_datetime, unit='ms')
    if resolution is None and len(data_types) > 1:
        try:
            index = collected[data_types[0]].index
            resolution = to_dt(index[1]) - to_dt(index[0])
        except IndexError:
            raise ValueError('No data for first type and no resolution given')

    new_index = None
    if resolution is not None:
        # np.datetime64 cannot hold timezone
        start = self.start_time.astimezone(timezone.utc).replace(tzinfo=None)
        end = self.end_time.astimezone(timezone.utc).replace(tzinfo=None)
        new_index = pd.date_range(start=pd.to_datetime(start).ceil(resolution),
                                  end=pd.to_datetime(end).ceil(resolution),
                                  freq=resolution)

    type_defs = self._client.get_types()

    converted: List[pd.DataFrame] = []
    for dt in data_types:
        if len(collected[dt].index) == 0:
            continue

        index = pd.DatetimeIndex(to_dt(collected[dt].index))
        data = collected[dt].data
        columns = [f'{dt} {field.name}' for field in type_defs[dt].fields]
        if include_can_source and dt[0].isalpha():
            columns = ['source'] + columns
        del collected[dt]  # allow GC of the data during loop
        df = pd.DataFrame.from_records(data=data,
                                       index=index,
                                       columns=columns)

        if not index.is_unique:
            # https://csrf.atlassian.net/browse/PLAT-222
            df = df[~index.duplicated()]
            index = df.index

        if 'source' in df.columns:
            out = []
            grouped = df.groupby('source', sort=False)
            for source, idx in grouped.groups.items():
                out.append(df.loc[idx, :]
                           .drop(columns=['source'])
                           .rename(mapper=lambda x: f'{x} {source}', axis=1))
        else:
            out = (df,)

        for df in out:
            if new_index is not None:
                df = df \
                    .reindex(index.union(new_index)) \
                    .interpolate(method='index') \
                    .reindex(new_index)

            if not df.empty:
                converted.append(df)

    if len(converted) == 0:
        return pd.DataFrame()
    else:
        return reduce(cast(DF_BiFunc, pd.DataFrame.join), converted)


def get_gradient_frame(self: Leg, **kwargs) -> pd.DataFrame:
    """
    Return computed gradient data as a time-series :code:`DataFrame`.

    :param generate: Request and wait for generation if not available
    :param timeout: How long to wait if requesting generation

    .. versionadded:: 1.4
    """
    type_defs = self._client.get_types()
    columns = ['9 ' + field.name for field in type_defs['9'].fields]

    data = self.get_gradient(**kwargs)
    if not data:
        return pd.DataFrame()

    # TODO: load csv directly?
    df = pd.DataFrame.from_records(
        ((pd.to_datetime(int(m.timestamp), unit='ms'), *(
            pd.to_numeric(v) for v in m.data.split(','))) for m in data),
        index='timestamp',
        columns=['timestamp', *columns]
    )
    return df


def get_transaction_frame(self: ChargerTransaction) -> pd.DataFrame:
    """
    Return meter readings as a time-series :code:`DataFrame`.

    .. versionadded:: 2.1
    """
    response = self._client.get(self.uri + '/data',
                                headers={'Accept': 'text/csv'})
    response.raise_for_status()
    df = pd.read_csv(io.StringIO(response.text),
                     names=['timestamp', 'value'],
                     dtype={'timestamp': 'int64',
                            'value': 'float64'},
                     index_col='timestamp')
    df.index = pd.to_datetime(df.index, unit='ms')
    return df


Leg.get_data_frame = get_data_frame
Trip.get_data_frame = get_data_frame
Leg.get_gradient_frame = get_gradient_frame
ChargerTransaction.get_data_frame = get_transaction_frame
