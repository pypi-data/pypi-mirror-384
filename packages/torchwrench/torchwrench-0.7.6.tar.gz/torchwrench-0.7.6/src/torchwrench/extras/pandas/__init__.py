#!/usr/bin/env python
# -*- coding: utf-8 -*-

from torchwrench.core.packaging import _PANDAS_AVAILABLE

if not _PANDAS_AVAILABLE:
    from torchwrench.extras.pandas import _pandas_fallback as pd

else:
    import pandas as pd


__all__ = [
    "_PANDAS_AVAILABLE",
    "pd",
]
