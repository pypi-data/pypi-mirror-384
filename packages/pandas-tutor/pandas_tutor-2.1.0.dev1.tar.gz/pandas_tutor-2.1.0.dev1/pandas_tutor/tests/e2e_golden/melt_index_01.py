# https://pandas.pydata.org/docs/reference/api/pandas.melt.html
import pandas as pd

df = pd.DataFrame({'A': {0: 'a', 1: 'b', 2: 'c'},
                   'B': {0: 1, 1: 3, 2: 5},
                   'C': {0: 2, 1: 4, 2: 6}})

df = df.set_index('A')

# tests that marks are right when left df index isn't default
df.melt()

