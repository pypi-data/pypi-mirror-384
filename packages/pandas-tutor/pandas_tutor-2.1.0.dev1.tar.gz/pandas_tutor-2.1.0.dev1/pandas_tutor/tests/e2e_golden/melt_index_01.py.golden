{
  "code": "# https://pandas.pydata.org/docs/reference/api/pandas.melt.html\nimport pandas as pd\n\ndf = pd.DataFrame({'A': {0: 'a', 1: 'b', 2: 'c'},\n                   'B': {0: 1, 1: 3, 2: 5},\n                   'C': {0: 2, 1: 4, 2: 6}})\n\ndf = df.set_index('A')\n\n# tests that marks are right when left df index isn't default\ndf.melt()\n\n",
  "explanation": [
    {
      "type": "MeltCall",
      "code_step": "df.melt()",
      "fragment": {
        "start": {
          "line": 0,
          "ch": 2
        },
        "end": {
          "line": 0,
          "ch": 9
        }
      },
      "marks": [
        {
          "type": "map_set",
          "maps": [
            {
              "type": "map",
              "from": {
                "type": "cell",
                "anchor": "lhs",
                "row": "a",
                "column": "B"
              },
              "to": {
                "type": "cell",
                "anchor": "rhs",
                "row": 0,
                "column": "variable"
              }
            },
            {
              "type": "map",
              "from": {
                "type": "cell",
                "anchor": "lhs",
                "row": "a",
                "column": "B"
              },
              "to": {
                "type": "cell",
                "anchor": "rhs",
                "row": 0,
                "column": "value"
              }
            },
            {
              "type": "map",
              "from": {
                "type": "cell",
                "anchor": "lhs",
                "row": "b",
                "column": "B"
              },
              "to": {
                "type": "cell",
                "anchor": "rhs",
                "row": 1,
                "column": "variable"
              }
            },
            {
              "type": "map",
              "from": {
                "type": "cell",
                "anchor": "lhs",
                "row": "b",
                "column": "B"
              },
              "to": {
                "type": "cell",
                "anchor": "rhs",
                "row": 1,
                "column": "value"
              }
            },
            {
              "type": "map",
              "from": {
                "type": "cell",
                "anchor": "lhs",
                "row": "c",
                "column": "B"
              },
              "to": {
                "type": "cell",
                "anchor": "rhs",
                "row": 2,
                "column": "variable"
              }
            },
            {
              "type": "map",
              "from": {
                "type": "cell",
                "anchor": "lhs",
                "row": "c",
                "column": "B"
              },
              "to": {
                "type": "cell",
                "anchor": "rhs",
                "row": 2,
                "column": "value"
              }
            }
          ]
        },
        {
          "type": "map_set",
          "maps": [
            {
              "type": "map",
              "from": {
                "type": "cell",
                "anchor": "lhs",
                "row": "a",
                "column": "C"
              },
              "to": {
                "type": "cell",
                "anchor": "rhs",
                "row": 3,
                "column": "variable"
              }
            },
            {
              "type": "map",
              "from": {
                "type": "cell",
                "anchor": "lhs",
                "row": "a",
                "column": "C"
              },
              "to": {
                "type": "cell",
                "anchor": "rhs",
                "row": 3,
                "column": "value"
              }
            },
            {
              "type": "map",
              "from": {
                "type": "cell",
                "anchor": "lhs",
                "row": "b",
                "column": "C"
              },
              "to": {
                "type": "cell",
                "anchor": "rhs",
                "row": 4,
                "column": "variable"
              }
            },
            {
              "type": "map",
              "from": {
                "type": "cell",
                "anchor": "lhs",
                "row": "b",
                "column": "C"
              },
              "to": {
                "type": "cell",
                "anchor": "rhs",
                "row": 4,
                "column": "value"
              }
            },
            {
              "type": "map",
              "from": {
                "type": "cell",
                "anchor": "lhs",
                "row": "c",
                "column": "C"
              },
              "to": {
                "type": "cell",
                "anchor": "rhs",
                "row": 5,
                "column": "variable"
              }
            },
            {
              "type": "map",
              "from": {
                "type": "cell",
                "anchor": "lhs",
                "row": "c",
                "column": "C"
              },
              "to": {
                "type": "cell",
                "anchor": "rhs",
                "row": 5,
                "column": "value"
              }
            }
          ]
        }
      ],
      "data": {
        "lhs": {
          "type": "DataFrame",
          "columns": {
            "names": [
              null
            ],
            "labels": [
              "B",
              "C"
            ]
          },
          "index": {
            "names": [
              "A"
            ],
            "labels": [
              "a",
              "b",
              "c"
            ]
          },
          "data": [
            [
              1,
              2
            ],
            [
              3,
              4
            ],
            [
              5,
              6
            ]
          ]
        },
        "rhs": {
          "type": "DataFrame",
          "columns": {
            "names": [
              null
            ],
            "labels": [
              "variable",
              "value"
            ]
          },
          "index": {
            "names": [
              null
            ],
            "labels": [
              0,
              1,
              2,
              3,
              4,
              5
            ]
          },
          "data": [
            [
              "B",
              1
            ],
            [
              "B",
              3
            ],
            [
              "B",
              5
            ],
            [
              "C",
              2
            ],
            [
              "C",
              4
            ],
            [
              "C",
              6
            ]
          ]
        }
      }
    }
  ]
}